libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt

```diff
@@ -28,37 +25,37 @@ open class YamuxHandler(
     initiator: Boolean
 ) : MuxHandler(ready, inboundStreamHandler) {
     private val idGenerator = AtomicInteger(if (initiator) 1 else 2) // 0 is reserved
-    private val receiveWindows = ConcurrentHashMap<MuxId, AtomicInteger>()
-    private val sendWindows = ConcurrentHashMap<MuxId, AtomicInteger>()
+    private val windowSizes = ConcurrentHashMap<MuxId, AtomicInteger>()
     private val sendBuffers = ConcurrentHashMap<MuxId, SendBuffer>()
-    private val totalBufferedWrites = AtomicInteger()
 
-    inner class SendBuffer(val ctx: ChannelHandlerContext) {
-        private val buffered = ArrayDeque<ByteBuf>()
+    inner class SendBuffer(val id: MuxId, val ctx: ChannelHandlerContext) {
+        private val bufferedData = ArrayDeque<ByteBuf>()
 
         fun add(data: ByteBuf) {
-            buffered.add(data.retain())
+            bufferedData.add(data)
```

Comment:
Right! It was excessive pair of retain/release. 
Just to note: the general rule of sending the handler closest to the wire is responsible for releasing the message buffer